WooCommerce/Classes/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorDataSource.swift

```diff
@@ -0,0 +1,35 @@
+import Yosemite
+
+/// `BottomSheetListSelectorDataSource` for selecting a sort order for the Products tab.
+///
+struct ProductsSortOrderBottomSheetListSelectorDataSource: BottomSheetListSelectorDataSource {
```

Comment:
Also, using `class` will protect us better since we can change the [`dataSource` property](https://github.com/woocommerce/woocommerce-ios/blob/4bd90f8e2a4464649085bd7e8ac9183a8a0f5e6f/WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift#L8) to a `private let`. I believe that's a really big win in terms of safety.

Reply:
great question! I used to use a `class` when there are states and access Objective-C API. I had a [PR comment](https://github.com/woocommerce/woocommerce-ios/pull/1430#discussion_r342928055) about why not using `struct` for `KeyboardFrameObserver`, and that's when I started making things `struct` by default unless it has a specific life cycle like `ProductImageActionHandler`.

aside from the PR feedback before, personally I prefer using `class` when there are mutable states, has a specific life cycle, or has blocks with self usage. for stateless helpers or data classes, I'd go with a `struct`.

@pmusolino how do you decide when to use `struct` or `class` when both work? maybe we can make this to our list of "dev preferences guideline doc" along with the xib/Storyboard/UI code guideline?

> Also, using class will protect us better since we can change the dataSource property to a private let. I believe that's a really big win in terms of safety.

yes definitely! I noticed this `var`, and went back to change it to `let` before then realized it's from the mutable `struct`. I'm going to change this `struct` to `class` for this reason!

Reply:
updated here: https://github.com/woocommerce/woocommerce-ios/pull/2156/commits/a24f87dd54299e4035206ceadd4f52d9534fc32a

Reply:
> We'd need to define "stateless" helpers though. 😄

haha yea - I see a class/struct as stateless if there are no mutable properties, and the properties have no mutable properties down the dependency hierarchy. how do you see this?

> From what I observed from myself throughout the years is that I started defaulting to struct because it's cool and new. But I realized that I was wrong. There's nothing wrong with using class. We shouldn't avoid it like the plague. And I don't think it's even expensive at all. Following Swift guidelines about using struct for "value types" makes more sense. If the requirements don't fit that, or the struct becomes more complex, it's fine to use class.

I agree to all of this 👍 I'd say let's follow the guidelines about using struct for "value types" like the models in the Networking layer. I vaguely remember some performance issues with using a `struct`, but it might have been improved in the recent Swift versions.

> KeyboardFrameObserver should definitely have been a class since it's managing state and even has NSNotification observation which should be cleared on deinit.

yes I think `KeyboardFrameObserver` should have been a class. when I was debugging the crash on `testObservingKeyboardFrameChangesWithoutKeyboardUserInfo` when running tests locally in https://github.com/woocommerce/woocommerce-ios/issues/1687, I was suspecting `KeyboardFrameObserver` wasn't deallocated correctly since it always crashed in the same test case. however, I wasn't able to reproduce the crash consistently so I couldn't verify the fix. I think it's worth converting `KeyboardFrameObserver` from `struct` to `class` in any case, just created an issue for this: https://github.com/woocommerce/woocommerce-ios/issues/2173